Create snooze task endpoint

joelclimbsthings opened the pull request:
Fixes #7368

**Add Endpoint**:
`POST - /wc-admin/onboarding/tasks/<task_id>/snooze`

**Parameters:**
- `duration`: Optional, currently supports `day|hour|week`, and defaults to `day`
- `task_list_id`: Optional, constrains search to specific task list

**Response**
Currently returns an error if the task does not exist, or the modified task definition. 


**Notes:**
- Also adding `isSnoozed` and `snoozedUntil` properties to the task definitions returned by `GET /wc-admin/onboarding/tasks`
- ~~Haven't implemented tracks yet (I'll be able to get this done early Monday)~~

### Detailed test instructions:

-   Checkout branch
-   Use a tools such as PostMan to send a request to `POST - /wc-admin/onboarding/tasks/invalid-task/snooze`
- Confirm that an error response is returned
- Now take a valid task ID and do the same request to `POST - /wc-admin/onboarding/tasks/<task_id>/snooze`
- Ensure that it returns the modified task definition with valid `isSnoozed` and `snoozedUntil` properties.
- Make a request to `GET - /wc-admin/onboarding/tasks` and check the definition of that task to ensure those properties are present and valid.

joelclimbsthings committed:
Adding basic snooze endpoint

joelclimbsthings committed:
Adding snooze props to tasks list and correcting snoozeUntil format

joelclimbsthings committed:
Adding changlog

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -36,6 +36,17 @@ class OnboardingTasks extends \WC_REST_Data_Controller {
+	 * @var string
+	 */
+	protected $duration_to_ms = array(
+		'day'  => 24 * 60 * 60 * 1000,
```
We have constants like `DAY_IN_SECONDS` that may be useful here instead of a new variable.

joelclimbsthings replied:
So far I decided to keep the mapping since I find it useful in accessing this in this class, but I'm now utilizing these constants. Thanks for the heads up 👍🏻

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -715,170 +767,252 @@ function( $has_installed, $bundle ) {
 
-		return rest_ensure_response( apply_filters( 'woocommerce_admin_onboarding_tasks', $tasks ) );
+		$snoozed_tasks     = get_option( 'woocommerce_task_list_remind_me_later_tasks' );
+		$tasks_with_snooze = array_map(
```
I think this could be added via filter to keep the logic down in this method.  See https://github.com/woocommerce/woocommerce-admin/pull/7538

joelclimbsthings replied:
Good call, done in 22297e280f1b810ab4484050a54119574913e499 👍🏻

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -715,170 +767,252 @@ function( $has_installed, $bundle ) {
+	 *
+	 * @return Object
+	 */
+	protected function get_task_by_id( $task_id ) {
```
Nice function!  In https://github.com/woocommerce/woocommerce-admin/pull/7538 I moved the tasks to the `OnboardingTasks` data store.  This method might also be more appropriate there.

joelclimbsthings replied:
I like it, done.

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -715,170 +767,252 @@ function( $has_installed, $bundle ) {
+	 */
+	protected function get_task_by_id( $task_id ) {
+		$tasks_res = $this->get_tasks()->data;
+		$task_list = (object) $tasks_res[0];
```
This will only get the first task list, which could create confusion.  Since we can have multiple task lists, I think we'll need to loop over all of them.

joelclimbsthings replied:
Good call, fixed 👍🏻

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -715,170 +767,252 @@ function( $has_installed, $bundle ) {
+		$tasks_with_snooze = array_map(
+			function ( $task ) use ( $snoozed_tasks ) {
+				$is_snoozed        = in_array( $task['id'], array_keys( $snoozed_tasks ), true );
+				$task['isSnoozed'] = $is_snoozed;
```
Should this check to see if the snooze time has already passed?

joelclimbsthings replied:
Perhaps, I'd assumed this option was already being managed via a CRON operation, but I'm glad to address this. I'm not sure what action we'd want to take in this case?

joshuatf replied:
I don't think this needs to be managed by cron.  If we store a timestamp with the date a task is snoozed until, we can compare that we when retrieve timestamps and check wether or not they're still snoozed:

```php
$task[ 'isSnoozed' ] = $snoozed_until > now();
```

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -715,170 +767,252 @@ function( $has_installed, $bundle ) {
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function snooze_task( $request ) {
```
I think we may need a property `isSnoozeable` in here (previously `allowRemindMeLater`) to determine whether a task is allowed to be snoozed or not.

joelclimbsthings replied:
Right, this is now implemented.

joshuatf reviewed (commented):
Thanks for getting this endpoint up and the extra utility functions!  Left a few comments in the code.

joelclimbsthings committed:
Restoring fuctionality after rebase, reworking get_task  utilities

joelclimbsthings committed:
Replacing duration properties with constants

joelclimbsthings committed:
Adding tracks to /snooze endpoint

joelclimbsthings committed:
Using utility function to get task for dismissable endpoint

joelclimbsthings committed:
Adding snooze status to tasks API via filter

joelclimbsthings committed:
Adding isSnoozeable to tasks and only allowing snooze when true

joelclimbsthings commented:
Thanks for the review @joshuatf ! I believe I've addressed your comments with the only exception being if some action should take place if the `snoozeUntil` timestamp has already passed. Any thoughts on what should occur in this case? 

Quick notes:
- Also added a optional param `task_list_id` to the `/snooze` endpoint as an optimization, and to avoid potential id conflicts across lists. 
- I noticed that the current logic supports undoing a snooze, although this wasn't mentioned in the original issue. Glad to add this; perhaps in a follow-up PR?

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -686,14 +742,10 @@ public function dismiss_task( $request ) {
-					break;
-				}
-			}
+		$dismiss_task = OnboardingTasksFeature::get_task_by_id( $id );
```
Thanks for using your new util here!  Nitpicky, but this variable sounds like a verb.  Maybe just `$task`?

```suggestion
		$task = OnboardingTasksFeature::get_task_by_id( $id );
```

joshuatf commented on src/Features/OnboardingTasks.php:
```diff
@@ -735,4 +787,26 @@ public function add_task_dismissal( $task_lists ) {
+		foreach ( $task_lists as $task_list_key => $task_list ) {
+			foreach ( $task_list['tasks'] as $task_key => $task ) {
+				if ( isset( $task['isSnoozeable'] ) && in_array( $task['id'], array_keys( $snoozed_tasks ), true ) ) {
+					$task_lists[ $task_list_key ]['tasks'][ $task_key ]['isSnoozed']    = true;
```
This bool can be dynamically created.   I haven't tested this, but something along the lines of:

```suggestion
					$task_lists[ $task_list_key ]['tasks'][ $task_key ]['isSnoozed']    = $snoozed_tasks[ $task['id'] ] > time();
```

joelclimbsthings replied:
Ah, understand what you mean here now. Done in https://github.com/woocommerce/woocommerce-admin/pull/7539/commits/1979e575c0431127490c2ded6dffc98872bd22dc.

joshuatf reviewed (commented):
Thanks for the changes, @joelclimbsthings!  Left a comment on one more item to address before merge.

>Also added a optional param task_list_id to the /snooze endpoint as an optimization, and to avoid potential id conflicts across lists.

Sounds good!  I think the list of tasks and retrieval could benefit from a better API, but maybe we can address this at the end of the cycle if time allows or during cooldown.  In the meantime, this is a good change 👍 

>I noticed that the current logic supports undoing a snooze, although this wasn't mentioned in the original issue. Glad to add this; perhaps in a follow-up PR?

Ya, we can either keep it on this endpoint and allow a `duration` param of `none` or `null` to undo the snooze or we can create a new endpoint in a follow-up.  What do you think is best?

joelclimbsthings committed:
Checking to ensure task is still snoozed when returning value

joelclimbsthings commented:
> Ya, we can either keep it on this endpoint and allow a duration param of none or null to undo the snooze or we can create a new endpoint in a follow-up. What do you think is best?

I'd say the latter option aligns more with what I would expect if consuming the API. Should be a quick follow-up, and I'm glad to create the issue and get that done today.

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -137,6 +148,37 @@ public function register_routes() {
+			'/' . $this->rest_base . '/(?P<id>[a-z0-9_-]+)/snooze',
+			array(
+				'args'   => array(
+					'task_id'      => array(
```
This arg should be removed; we can use the `id` retrieved from the path.

joelclimbsthings replied:
Whoops, that must have snuck in when I rebased 🤭

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -736,4 +788,51 @@ public function undo_dismiss_task( $request ) {
+	 */
+	public function snooze_task( $request ) {
+
+		$task_id         = $request->get_param( 'task_id' );
```
```suggestion
		$task_id         = $request->get_param( 'id' );
```

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -736,4 +788,51 @@ public function undo_dismiss_task( $request ) {
+		if ( ! $is_snoozeable ) {
+			return new \WP_Error(
+				'woocommerce_tasks_invalid_task',
+				__( 'Sorry, no snoozeable task with that ID was found.', 'woocommerce-admin' )
```
We can also add a status code here.

```suggestion
				__( 'Sorry, no snoozeable task with that ID was found.', 'woocommerce-admin' ),
				array(
					'status' => 404,
				)
```

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -736,4 +788,51 @@ public function undo_dismiss_task( $request ) {
+			);
+		}
+
+		$snooze_option = get_option( 'woocommerce_task_list_remind_me_later_tasks' );
```
Needs default param to avoid error on initial snooze:

```suggestion
		$snooze_option = get_option( 'woocommerce_task_list_remind_me_later_tasks', array() );
```

joelclimbsthings replied:
Good catch 👍🏻

joshuatf commented on src/Features/OnboardingTasks.php:
```diff
@@ -716,6 +717,57 @@ function( $has_installed, $bundle ) {
+			return null;
+		}
+
+		$tasks_to_search = $task_list_id ? $task_list['tasks'] : array_reduce(
```
Not a blocker to getting this merged, especially since we may update the task API by the end of cycle, but just a quick note that this loops over the tasks twice.

This most likely will not have any meaningful impact on performance given how small the task list is, but I'd opt for a single loop where possible if we update this.

joelclimbsthings replied:
I don't believe this is really any different than the nested `foreach` loops we'd be using otherwise? The reduce only loops through the task lists (only 1 currently), and then the following `foreach` loop just loops through the relevant tasks. The benefit is that if `$task_list_id` is provided, you'll skip the first loop entirely and only loop through the tasks for the specified list. 

This depends a bit on the performance of `array_merge`; is that what you're referring to? Not too familiar with the inner-workings there.

joshuatf replied:
Yes and no.  You're correct if we simply nest the foreach and let it run.  However, with the nested option we can return early or break the loop which has the ability to be more performant.

But again, this is a very nitpicky comment and I'm hoping we can create a better API around registering/getting tasks in the backend that would make this obsolete anyway.

joshuatf replied:
Yes and no.  You're correct if we simply nest the foreach and let it run.  However, with the nested option we can return early or break the loop which has the ability to be more performant.

But again, this is a very nitpicky comment and I'm hoping we can create a better API around registering/getting tasks in the backend that would make this obsolete anyway.

joelclimbsthings replied:
True, although the reduce is only looping through the list objects, and the subsequent `foreach` loop that goes through the tasks does return early as you mention. Hopefully all moot as we improve this, but I was just curious if I was missing something. 🙏

joshuatf reviewed (approved):
Looks good, Joel!  Encountered a few minor errors and left suggestions on fixes, but pre-approving pending those fixes.

Thanks for the work on this 💯

joelclimbsthings committed:
Removing task_id arg definition

joelclimbsthings committed:
Adding status and default array for options

joelclimbsthings committed:
Minor tweak to spacing
